crates/scalajs/scala/wit.scala

```diff
@@ -0,0 +1,685 @@
+package object wit {
+
+  import scala.scalajs.js
+  import scala.scalajs.js.JSConverters._
+  import scala.scalajs.js.|
+  import scala.scalajs.js.annotation.JSName
+
+  sealed trait Nullable[+A] extends js.Any
+
+  object Nullable {
+    def some[A](value: A): Nullable[A] = value.asInstanceOf[Nullable[A]]
+
+    val none: Nullable[Nothing] = null.asInstanceOf[Nullable[Nothing]]
+
+    def fromOption[A](option: Option[A]): Nullable[A] =
+      option match {
+        case Some(value) => some(value)
+        case None => none
+      }
+  }
+
+  implicit class NullableOps[A](private val self: Nullable[A]) extends AnyVal {
+    def toOption: Option[A] = Option(self.asInstanceOf[A])
+  }
+
+  sealed trait WitOption[A] extends js.Object {
+    val tag: String
+    val `val`: js.UndefOr[A]
+  }
+
+  object WitOption {
+    def some[A](value: A): WitOption[A] = new WitOption[A] {
+      val tag: String = "some"
+      val `val`: js.UndefOr[A] = value
+    }
+
+    def none[A]: WitOption[A] = new WitOption[A] {
+      val tag: String = "none"
+      val `val`: js.UndefOr[A] = js.undefined
+    }
+
+    def fromOption[A](option: Option[A]): WitOption[A] =
+      option match {
+        case Some(value) => some(value)
+        case None => none
+      }
+  }
+
+  implicit class WitOptionOps[A](private val self: WitOption[A]) extends AnyVal {
+    def toOption: Option[A] = self.tag match {
+      case "some" => Some(self.`val`.get)
+      case _ => None
+    }
+  }
+
+  sealed trait WitResult[Ok, Err] extends js.Object {
+    val tag: String
+    val `val`: js.UndefOr[Ok | Err]
+  }
+
+  object WitResult {
+    def ok[Ok, Err](value: Ok): WitResult[Ok, Err] = new WitResult[Ok, Err] {
+      val tag: String = "ok"
+      val `val`: js.UndefOr[Ok | Err] = value
+    }
+
+    def err[Ok, Err](value: Err): WitResult[Ok, Err] = new WitResult[Ok, Err] {
+      val tag: String = "err"
+      val `val`: js.UndefOr[Ok | Err] = value
+    }
+
+    def fromEither[E, A](either: Either[E, A]): WitResult[A, E] =
+      either match {
+        case Right(value) => ok(value)
+        case Left(value) => err(value)
+      }
+  }
+
+  implicit class WitResultOps[Ok, Err](private val self: WitResult[Ok, Err]) extends AnyVal {
+    def toEither: Either[Err, Ok] = self.tag match {
+      case "ok" => Right(self.`val`.get.asInstanceOf[Ok])
+      case _ => Left(self.`val`.get.asInstanceOf[Err])
+    }
+  }
+
+  type WitList[A] = js.Array[A]
+
+  object WitList {
+    def fromList[A](list: List[A]): WitList[A] = list.toJSArray
+  }
+
+  sealed trait WitTuple0 extends js.Object {
+  }
+
+  object WitTuple0 {
+    def apply(): WitTuple0 = js.Array().asInstanceOf[WitTuple0]
+
+    def unapply(tuple: WitTuple0): Some[Unit] = Some(())
+
+    implicit def fromScalaTuple0(tuple: Unit): WitTuple0 = WitTuple0()
+
+    implicit def toScalaTuple0(tuple: WitTuple0): Unit = ()
+  }
+
+  sealed trait WitTuple1[T1] extends js.Object {
```

Comment:
Looks like `WitTupleN[...]` could be an alias of `js.TupleN[...]`.

Reply:
Yeah I've copied that from `js.Tuple2` but found there is only `Tuple2` and `Tuple3` in the scala.js library while you can definitely use bigger arities in WIT.

Reply:
The other ones are here: https://github.com/scala-js/scala-js/blob/main/library/src/main/scala/scala/scalajs/js/Tuple.nodoc.scala so that they don't clutter the Scaladoc. But they do exist ;)

Reply:
Thanks, then these types are really not needed. Will remove them.